wire-ios/Wire-iOS/Sources/URLActionRouter.swift

```diff
@@ -202,37 +204,46 @@ extension URLActionRouter: PresentationDelegate {
         presentAlert(alert)
     }
 
-    private func presentCustomBackendAlert(with configurationURL: URL) {
-        let alert = UIAlertController(title: L10n.Localizable.UrlAction.SwitchBackend.title,
-                                      message: L10n.Localizable.UrlAction.SwitchBackend.message(configurationURL.absoluteString),
-                                      preferredStyle: .alert)
-
-        let agreeAction = UIAlertAction(title: L10n.Localizable.General.ok, style: .default) { [weak self] _ in
-            self?.rootViewController.isLoadingViewVisible = true
-            self?.switchBackend(with: configurationURL)
+    private func switchBackend(configURL: URL) {
+        guard
+            SecurityFlags.customBackend.isEnabled,
+            let sessionManager
+        else {
+            return
         }
-        alert.addAction(agreeAction)
-
-        let cancelAction = UIAlertAction(title: L10n.Localizable.General.cancel, style: .cancel)
-        alert.addAction(cancelAction)
 
-        presentAlert(alert)
-    }
+        sessionManager.fetchBackendEnvironment(at: configURL) { [weak self] result in
+            guard let self else { return }
 
-    private func switchBackend(with configurationURL: URL) {
-        sessionManager?.switchBackend(configuration: configurationURL) { [weak self] result in
-            self?.rootViewController.isLoadingViewVisible = false
             switch result {
-            case let .success(environment):
-                BackendEnvironment.shared = environment
-            case let .failure(error):
-                guard let strongSelf = self else { return }
-                let localizedError = strongSelf.mapToLocalizedError(error)
-                strongSelf.presentLocalizedErrorAlert(localizedError)
+            case .success(let backendEnvironment):
+                self.requestUserConfirmationToSwitchBackend(backendEnvironment) { didConfirm in
+                    guard didConfirm else { return }
+                    sessionManager.switchBackend(to: backendEnvironment)
+                    BackendEnvironment.shared = backendEnvironment
```

Comment:
I would expect that switchBackend take care of assigning `BackendEnvironment.shared = backendEnvironment` no?

Reply:
Yes me too, but `BackendEnvironment.shared` is defined in WireCommonComponents (to be accessible for share and notification extensions) and the `switchBackend` method is in WireSyncEngine. I tried to refactor it all so that we didn't have the singleton but the changes snowballed and I stopped. I'll create a ticket for the refactoring idea.

Reply:
good to know